// sequelize-typescript/lib/hooks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// annotations
var BeforeValidate_1 = require("./annotations/BeforeValidate");
exports.BeforeValidate = BeforeValidate_1.BeforeValidate;
var AfterValidate_1 = require("./annotations/AfterValidate");
exports.AfterValidate = AfterValidate_1.AfterValidate;
var ValidationFailed_1 = require("./annotations/ValidationFailed");
exports.ValidationFailed = ValidationFailed_1.ValidationFailed;
var BeforeCreate_1 = require("./annotations/BeforeCreate");
exports.BeforeCreate = BeforeCreate_1.BeforeCreate;
var AfterCreate_1 = require("./annotations/AfterCreate");
exports.AfterCreate = AfterCreate_1.AfterCreate;
var BeforeDestroy_1 = require("./annotations/BeforeDestroy");
exports.BeforeDestroy = BeforeDestroy_1.BeforeDestroy;
var AfterDestroy_1 = require("./annotations/AfterDestroy");
exports.AfterDestroy = AfterDestroy_1.AfterDestroy;
var BeforeRestore_1 = require("./annotations/BeforeRestore");
exports.BeforeRestore = BeforeRestore_1.BeforeRestore;
var AfterRestore_1 = require("./annotations/AfterRestore");
exports.AfterRestore = AfterRestore_1.AfterRestore;
var BeforeUpdate_1 = require("./annotations/BeforeUpdate");
exports.BeforeUpdate = BeforeUpdate_1.BeforeUpdate;
var AfterUpdate_1 = require("./annotations/AfterUpdate");
exports.AfterUpdate = AfterUpdate_1.AfterUpdate;
var BeforeSave_1 = require("./annotations/BeforeSave");
exports.BeforeSave = BeforeSave_1.BeforeSave;
var AfterSave_1 = require("./annotations/AfterSave");
exports.AfterSave = AfterSave_1.AfterSave;
var BeforeUpsert_1 = require("./annotations/BeforeUpsert");
exports.BeforeUpsert = BeforeUpsert_1.BeforeUpsert;
var AfterUpsert_1 = require("./annotations/AfterUpsert");
exports.AfterUpsert = AfterUpsert_1.AfterUpsert;
var BeforeBulkCreate_1 = require("./annotations/BeforeBulkCreate");
exports.BeforeBulkCreate = BeforeBulkCreate_1.BeforeBulkCreate;
var AfterBulkCreate_1 = require("./annotations/AfterBulkCreate");
exports.AfterBulkCreate = AfterBulkCreate_1.AfterBulkCreate;
var BeforeBulkDestroy_1 = require("./annotations/BeforeBulkDestroy");
exports.BeforeBulkDestroy = BeforeBulkDestroy_1.BeforeBulkDestroy;
var AfterBulkDestroy_1 = require("./annotations/AfterBulkDestroy");
exports.AfterBulkDestroy = AfterBulkDestroy_1.AfterBulkDestroy;
var BeforeBulkRestore_1 = require("./annotations/BeforeBulkRestore");
exports.BeforeBulkRestore = BeforeBulkRestore_1.BeforeBulkRestore;
var AfterBulkRestore_1 = require("./annotations/AfterBulkRestore");
exports.AfterBulkRestore = AfterBulkRestore_1.AfterBulkRestore;
var BeforeBulkUpdate_1 = require("./annotations/BeforeBulkUpdate");
exports.BeforeBulkUpdate = BeforeBulkUpdate_1.BeforeBulkUpdate;
var AfterBulkUpdate_1 = require("./annotations/AfterBulkUpdate");
exports.AfterBulkUpdate = AfterBulkUpdate_1.AfterBulkUpdate;
var BeforeFind_1 = require("./annotations/BeforeFind");
exports.BeforeFind = BeforeFind_1.BeforeFind;
var BeforeFindAfterExpandIncludeAll_1 = require("./annotations/BeforeFindAfterExpandIncludeAll");
exports.BeforeFindAfterExpandIncludeAll = BeforeFindAfterExpandIncludeAll_1.BeforeFindAfterExpandIncludeAll;
var BeforeFindAfterOptions_1 = require("./annotations/BeforeFindAfterOptions");
exports.BeforeFindAfterOptions = BeforeFindAfterOptions_1.BeforeFindAfterOptions;
var AfterFind_1 = require("./annotations/AfterFind");
exports.AfterFind = AfterFind_1.AfterFind;
var BeforeCount_1 = require("./annotations/BeforeCount");
exports.BeforeCount = BeforeCount_1.BeforeCount;
var BeforeDelete_1 = require("./annotations/BeforeDelete");
exports.BeforeDelete = BeforeDelete_1.BeforeDelete;
var AfterDelete_1 = require("./annotations/AfterDelete");
exports.AfterDelete = AfterDelete_1.AfterDelete;
var BeforeBulkDelete_1 = require("./annotations/BeforeBulkDelete");
exports.BeforeBulkDelete = BeforeBulkDelete_1.BeforeBulkDelete;
var AfterBulkDelete_1 = require("./annotations/AfterBulkDelete");
exports.AfterBulkDelete = AfterBulkDelete_1.AfterBulkDelete;
